internal/remotewrite/proxy.go

```diff
@@ -57,41 +59,72 @@ func remoteWrite(write *url.URL, endpoints []Endpoint, logger log.Logger) http.H
+					promconfig.WithDialContextFunc((&net.Dialer{
+						Timeout:   30 * time.Second,
+						KeepAlive: 30 * time.Second,
+					}).DialContext))
 				if err != nil {
-					level.Error(rlogger).Log("failed to create a new HTTP client", "err", err)
+					//level.Error(rlogger).Log("msg", "failed to create a new HTTP client", "err", err)
+					LogChannels[i] <- logMessage{
+						messageKey: "failed to create a new HTTP client",
+						keyvals: []interface{}{
+							"msg", "failed to create a new HTTP client", "err", err,
+						}}
 				}
 			}
 
 			req, err := http.NewRequest(http.MethodPost, endpoint.URL, bytes.NewReader(body))
 			req.Header = r.Header
 			if err != nil {
-				level.Error(rlogger).Log("msg", "Failed to create the forward request", "err", err, "url", endpoint.URL)
+				//level.Error(rlogger).Log("msg", "Failed to create the forward request", "err", err, "url", endpoint.URL)
+				LogChannels[i] <- logMessage{
+					messageKey: "Failed to create the forward request",
+					keyvals: []interface{}{
+						"msg", "Failed to create the forward request", "err", err,
+					}}
 			} else {
 				ep := endpoint
+				j := i
 				go func() {
 					resp, err := client.Do(req)
 					if err != nil {
 						remotewriteRequests.With(prometheus.Labels{"code": "<error>", "name": ep.Name}).Inc()
-						level.Error(rlogger).Log("msg", "Failed to send request to the server", "err", err)
+						//level.Error(rlogger).Log("msg", "Failed to send request to the server", "err", err)
+						LogChannels[j] <- logMessage{
+							messageKey: "Failed to send request to the server",
+							keyvals: []interface{}{
+								"msg", "Failed to send request to the server", "err", err,
+							}}
 					} else {
 						defer resp.Body.Close()
 						remotewriteRequests.With(prometheus.Labels{"code": strconv.Itoa(resp.StatusCode), "name": ep.Name}).Inc()
 						if resp.StatusCode >= 300 || resp.StatusCode < 200 {
 							responseBody, err := ioutil.ReadAll(resp.Body)
 							if err != nil {
-								level.Error(rlogger).Log("msg", "Failed to read response of the forward request", "err", err, "return code", resp.Status, "url", ep.URL)
+								//level.Error(rlogger).Log("msg", "Failed to read response of the forward request", "err", err, "return code", resp.Status, "url", ep.URL)
+								LogChannels[j] <- logMessage{
+									messageKey: "Failed to forward metrics" + resp.Status,
+									keyvals: []interface{}{
+										"msg", "Failed to forward metrics", "return code", resp.Status, "url", ep.URL,
+									}}
 							} else {
-								level.Error(rlogger).Log("msg", "Failed to forward metrics", "return code", resp.Status, "response", string(responseBody), "url", ep.URL)
+								LogChannels[j] <- logMessage{
+									messageKey: "Failed to forward metrics" + resp.Status,
+									keyvals: []interface{}{
+										"msg", "Failed to forward metrics", "return code", resp.Status, "response", string(responseBody), "url", ep.URL}}
```

Comment:
What is purpose to handle `err != nil` and `err == nil` separately?

Reply:
err == nil means the response body can be parsed, then we can log it out for better assist in PD